### PR TITLE
86 application window

### LINKED ISSUE #86: Application window
## Reason
To be able to render the application, it has to have a window in which to place rendered frames.

## Requirements
- [x] When rendering is enabled, an application window is shown 
- [x] `fn main` is located in the client code (just like our current example tests), _not_ in the graphics library
- [x] The rendering feature is _optional_ and an application can run entirely without it (just like before this feature was implemented) (there should be a feature that enables compilation of this, and it should be possible to compile without)
- [x] Rendering of the application window does not prevent nor impede the ECS from running

## Dependencies
- [x] #82

### PR DESCRIPTION
**Pull request by martinjonsson01**
Enables rendering of an application window alongside the ECS running in the background. Part of this meant we had to generalize how plugins are added to the ECS engine. 

> Note: this is a sub-task of #61.

The window is currently completely empty, as rendering of contents will be implemented in #87.
image

The new way of adding a plugin is:
```rust
// Define a trait which describes a property which the application can have:
pub trait Audible {
    fn with_audio(self) -> AudioResult<()>;
}

// Implement the trait for all types of application:
impl<App: Application> Audible for App {
...
}
```
(see `gfx-plugin\lib.rs` for a real example of this)

and now this plugin can be enabled by calling:
```rust
let mut app = BasicApplication::default().with_audio()?;
```

Closes #86

### CODE REVIEW

File: crates/gfx-plugin/examples/rendered-ecs.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+use scheduler::schedule::PrecedenceGraph;
+use std::thread;
+use std::time::Duration;
+use tracing::{instrument, warn};
+
+// a simple example of how to use the crate `ecs`
+#[instrument]
+fn main() -> Result<(), Report> {
+    let mut app = BasicApplication::default()
+        .with_rendering()?
```
Comment (EdvinNilsson) [inline crates/gfx-plugin/examples/rendered-ecs.rs:20]: It's nice that it's so easy to add the graphics plugin

File: crates/gfx-plugin/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
+//! A rendering-plugin for RECS, which renders the [`BasicApplication`] using a renderer
```
Comment (EdvinNilsson) [inline crates/gfx-plugin/src/lib.rs:1]:
```suggestion
//! A rendering-plugin for RECS, which renders the [`BasicApplication`](ecs::BasicApplication) using a renderer
```

File: crates/gfx/src/engine.rs [hunk 3]
```diff
@@ -100,94 +100,86 @@ pub enum MainMessage {
...
-        + Send
-        + Sync
-        + 'a,
 {
     /// Creates a new instance of `Engine`.
     ///
-    /// The `initialize_gfx` function is called once, before beginning the render loop.
-    /// The `simulate` function is called from the simulation thread every simulation tick.
+    /// [`GraphicsInitializer`] is called once, before beginning the render loop.
+    /// [`Simulation`] is called from the simulation thread every simulation tick.
```
Comment (EdvinNilsson) [inline crates/gfx/src/engine.rs:138]: Cannot find `GraphicsInitializer` and `Simulation` in the code.

File: crates/ecs/src/profiling.rs [hunk 4]
```diff
@@ -19,10 +19,15 @@ pub enum ProfilingError {
 /// Whether a profiling operation succeeded.
 pub type ProfilingResult<T, E = ProfilingError> = Result<T, E>;
 
-impl Application {
+/// Represents an [`Application`] which can be profiled.
+pub trait Profileable: Sized {
     /// Enables profiling connection to Tracy.
+    fn with_profiling(self) -> ProfilingResult<Self>;
```
Comment (JacobBredin) [inline crates/ecs/src/profiling.rs:25]: Does this pattern have a name? I'm really enjoying it 😏

**Review by EdvinNilsson — APPROVED** [review]
LGTM!

File: crates/gfx-plugin/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    /// Initializes and configures rendering functionality, enabling visualization of
+    /// the simulation.
+    fn with_rendering(self) -> GfxAppResult<RenderedApp>;
+}
+
+type GraphicsEngine = gfx::engine::Engine<NoUI, Vec<Object>>;
+type GraphicsEngineHandle = gfx::engine::EngineHandle<Vec<Object>>;
+
+impl<App: Application + Send> Graphical<GraphicalApplication<App>> for App {
+    fn with_rendering(self) -> GfxAppResult<GraphicalApplication<App>> {
```
Comment (JacobBredin) [inline crates/gfx-plugin/src/lib.rs:149]: I like the idea of "decorating" the app, makes the builder very clean 👍

**Review by JacobBredin — APPROVED** [review]
LGTM!

Cool Ideas used 👍

File: crates/gfx/src/engine.rs (hunk 3 above)
Comment (martinjonsson01) [reply crates/gfx/src/engine.rs:138]: Oops, that's old stuff. We'll remove it.

File: crates/ecs/src/profiling.rs (hunk 4 above)
Comment (martinjonsson01) [reply crates/ecs/src/profiling.rs:25]: I'm not sure if it has its own name, but it's inspired by how builders work!

**Pull request merged**